[RFC] audit plugin rework

jeffmahoney opened the pull request:
This is meant for discussion, not inclusion just yet.

chajain committed:
Extending audit Plugin and using it for creating audit artifacts

jeffmahoney committed:
audit: rearchitect to share rule handling among all invocations

The audit plugin doesn't share anything among invocations, which makes
sense given the historical model of preconfiguring of the rules,
subscribing the firehose, and sorting out the events later.  The problem
with that architecture is that if anything external happens with the
audit subsystem -- even as simple as restarting the audit daemon -- the
audit rules are lost and no more events will be generated.

By specifying which rules a particular artifact depends on in the
arguments, we can be certain that the rules required will always be in
effect, can log when they are not, and restore them automatically.

Since multiple invocations can ultimately require the same audit rules,
management of those rules must be centralized and reference counted.  (I
suppose an alternative is is just to let each invocation flail trying to
restore its own rules, but that's just messy.)

There are a few changes required to make this go:
1) We need a separate command-and-control audit client to do the
   configuration, checking, and fixing
2) We need a goroutine that periodically checks to ensure that the rules
   are correct.  This is implemented as something that sits idle waiting
   for configuration events and then batches them until 500ms has passed
   since receipt of the last one.  The reason for this is that e.g.
   "auditctl -D" will generate an event for every single rule removed
   and we don't need to check the rules every time.
3) Each invocation will receive a copy of each event and also log
   messages that apply to the invocation as needed.
4) The lifetime of the shared services is tied to the listeners.
   Once the last listener exits, the shared services shutdown.  If new
   invocations come in, a new instance will be started up.

jeffmahoney committed:
third_party/go-libaudit: rule.Rule.Build: Don't assume that no syscalls means all syscalls

Rule.Build assumes that if no syscalls are specified they all are set.
This is really only the case when the exit list is used since the
syscall numbers aren't available in the other lists.  When we assume
that all of the syscalls are enabled, we end up generating wireformat
rules for e.g. 'task,never' that have all of the syscall bits set.  That
doesn't match what is already used when 'auditctl -a task,never' is
used.  It may be ignored by the kernel when such a rule is added, but
it would cause problems when that rule is deleted.

jeffmahoney committed:
third_party/go-libaudit: Add support for removing individual rules

We currently don't handle the '-d' or '-W' options that would remove
list rules or file watches.  This commit adds support to handle those
properly.  rule.ToCommandLine still returns the expected result, but
I've added a rule.ToCommandLineAddRemove that takes a bool indicating
whether the rule would be added or removed.  This was required to do
testing of deletion rules.

jeffmahoney committed:
import go-libaudit to handle individual rule removal

[temporary import]

elastic/go-libaudit is used by Velociraptor to handle audit rules
changes in the kernel.  The upstream version doesn't have the capability
to remove individual rules.  I've submitted a PR to that project to
implement it, but we can't wait for it to be accepted.

jeffmahoney committed:
third_party/go-libaudit: Report missing rules during deletion

Although the implementation for AuditClient.Delete is only used to back
AuditClient.DeleteAll, we'd like to be able to delete individual rules.
This commit adds checking of the netlink error field and reports when
the deletion has failed.  When DeleteAll is called, we ignore the ENOENT
return since it could've raced somewhere and we don't actually care
since we're deleting all of the rules.

guangyee commented on artifacts/definitions/Linux/Events/ExecutableFiles.yaml:
```diff
@@ -0,0 +1,36 @@
+sources:
+  
+  - query: |
+     LET exec_bit_rules = ("-a always,exit -F arch=b64 -S fchmodat -F a2&73 -k vrr_exec_bit_addition", "-a always,exit -F arch=b32 -S fchmodat -F a2&73 -k vrr_exec_bit_addition", "-a always,exit -F arch=b64 -S fchmod -F a1&73 -k vrr_exec_bit_addition", "-a always,exit -F arch=b32 -S fchmod -F a1&73 -k vrr_exec_bit_addition")
```
This rule seem to be picking `/Null` as one of the changes. Not sure what caused it.

lorddoskias commented on vql/linux/audit.go:
```diff
@@ -37,15 +43,54 @@ func (self *streamHandler) outputMultipleMessages(msgs []*auparse.AuditMessage)
+                      s.Log("Error Building Rule: %s", err)
+                      return fmt.Errorf("Error while adding rules: %w", err)
+               }
+               defer c.GetRules()
```
Why do you do GetRules() in a deferred context? It simply returns all rules in  binary form i.e not doing any refcounting?

jeffmahoney replied:
This is part of the code that is removed in the subsequent commit

lorddoskias commented on vql/linux/audit.go:
```diff
@@ -4,90 +4,688 @@ package linux
+		if ok {
+			return fmt.Errorf("Cannot add duplicate rule `%s' in single query", line)
+		}
+
```
nit: this could be written as 

```
if _, ok := self.rules[watcherRule.rule]; ok { 
       return fmt.Errorf("Cannot add duplicate rule `%s' in single query", line)
}
```

jeffmahoney replied:
It could, but I really don't like that pattern.

lorddoskias commented on vql/linux/audit.go:
```diff
@@ -4,90 +4,688 @@ package linux
+	if !batchTimeout.Stop() {
+		<- batchTimeout.C
+	}
+
```
Why do you stop the timer here?

jeffmahoney replied:
Because there's no way to create a timer without also starting it, and it's simpler to create it and stop it than to create it in the loop once

lorddoskias replied:
Right, this actually is from the documentation on the golang page.

jeffmahoney replied:
I also ended up pulling that out into a helper so it's clear what the intent is

lorddoskias commented on vql/linux/audit.go:
```diff
@@ -4,90 +4,688 @@ package linux
+
+		err := auditService.setupWatcherService()
+		if err != nil {
+			auditService = nil
```
This is buggy, if setupWatcherService returns this function will return nil but at the end of the function you still return auditService, nil, so 'err' should be returned at the end of the function. On a kernel which doesn't support audit this leads to a null ptr deref: 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x15edc32]

goroutine 223 [running]:
www.velocidex.com/golang/velociraptor/vql/linux.(*AuditWatcherService).addRules(0x0, 0x40)
	/root/security-sensor/vql/linux/audit.go:466 +0x52
www.velocidex.com/golang/velociraptor/vql/linux.(*AuditWatcherService).connectWatcher(0x0, {0x2309ca0, 0xc001074000}, 0xc001188700)
	/root/security-sensor/vql/linux/audit.go:394 +0x3d
www.velocidex.com/golang/velociraptor/vql/linux.AuditPlugin.Call.func1()
	/root/security-sensor/vql/linux/audit.go:737 +0x3cb
created by www.velocidex.com/golang/velociraptor/vql/linux.AuditPlugin.Call
	/root/security-sensor/vql/linux/audit.go:697 +0xfd
```

jeffmahoney replied:
Fixed in the updated PR

lorddoskias commented on vql/linux/audit.go:
```diff
@@ -4,90 +4,688 @@ package linux
+	defer self.rulesLock.Unlock()
+
+	self.notifyWatchers("audit: detected audit configuration change")
+	self.logger.Info("audit: detected audit configuration change")
```
The call to self.logger.Info is redundant since notifyWatchers will cause every watched to print the message. This results in duplicated message printout: 

```
[INFO] 2022-06-10T10:54:41Z Compiled all artifacts. 
[INFO] 2022-06-10T10:54:41Z audit: detected audit configuration change 
[INFO] 2022-06-10T10:54:41Z audit: detected audit configuration change 
```

jeffmahoney replied:
Yep. It'll always result in duplicate message printout. In fact, there'll be a message for every watcher. IMO this is a problem with Velociraptor. If we don't log it multiple times, it won't appear in the log in the UI for the plugin. I get that you're saying that it is separate from the scope, but I don't think one more entry hurts when there will already be several. Eventually, I'd like to fix it so that the logging works differently.

lorddoskias commented on artifacts/definitions/Linux/Events/NewFiles.yaml:
```diff
@@ -0,0 +1,36 @@
+sources:
+  - query: |
+
+     LET new_file_rules = ("-w /etc/init.d -p wa -k vrr_etc_update", "-w /etc/rc.local -p wa -k vrr_etc_update", "-w /etc/modules -p wa -k vrr_etc_update", "-w /etc/cron.allow -p wa -k vrr_etc_update", "-w /etc/cron.deny -p wa -k vrr_etc_updae", "-w /etc/cron.d/ -p wa -k vrr_etc_update", "-w /etc/cron.daily/ -p wa -k vrr_etc_update", "-w /etc/cron.hourly/ -p wa -k vrr_etc_update", "-w /etc/cron.m ...
```
I see that in those rules there are the crontab directories. So when a crontab is entry is changed/removed the following output is produced: 

> 
> [INFO] 2022-06-10T11:00:39Z hash /var/spool/cron/crontabs/tmp.WStmBd: lstat /var/spool/cron/crontabs/tmp.WStmBd: no such file or directory 
> [INFO] 2022-06-10T11:00:39Z hash /var/spool/cron: Only regular files supported 
> [INFO] 2022-06-10T11:00:39Z hash /var/spool/cron/crontabs/tmp.WStmBd: lstat /var/spool/cron/crontabs/tmp.WStmBd: no such file or directory 
> [
>  {
>   "Time": "2022-06-10T11:00:38.726Z",
>   "UserId": 0,
>   "User": "root",
>   "State": "success",
>   "FileName": "crontabs/tmp.WStmBd",
>   "FilePath": "/var/spool/cron/crontabs/tmp.WStmBd",
>   "Hash_Sha256": null,
>   "Hash_Sha1": null
>  },
>  {
>   "Time": "2022-06-10T11:00:38.73Z",
>   "UserId": 0,
>   "User": "root",
>   "State": "success",
>   "FileName": "",
>   "FilePath": "/var/spool/cron/",
>   "Hash_Sha256": null,
>   "Hash_Sha1": null
>  },
>  {
>   "Time": "2022-06-10T11:00:38.73Z",
>   "UserId": 0,
>   "User": "root",
>   "State": "success",
>   "FileName": "crontabs/tmp.WStmBd",
>   "FilePath": "/var/spool/cron/crontabs/tmp.WStmBd",
>   "Hash_Sha256": null,
>   "Hash_Sha1": null
>  }
> ]

I think that's suboptimal because audit can't distinguish between the actual user spool file and the temporary files created by 'crontab -e'. This is being handled in the cronsnoop plugin I'm working on, so how do we want to deal with this - here in audit or in the cronsnoop plugin?

jeffmahoney replied:
If the cronsnoop plugin is smarter, let's use that.

guangyee commented on artifacts/definitions/Linux/Events/ProcessStatuses.yaml:
```diff
@@ -0,0 +1,40 @@
+sources:
+  - query: |
+
+     LET proc_stat_rules = ("-a always,exit -F arch=b64 -S execve -F euid!=0 -F auid!=-1 -k vrr_proexec", "-a always,exit -F arch=b32 -S execve -F euid!=0 -F auid!=-1 -k vrr_proexec")
```
This rule seem to log every single command executed by non-root users, even if the command with exist code other than zero (0). Moreover, in audit.log, it's showing the command as `success`  or `exit=0`, which does not seem to reflect the actual exit code of the command itself.

jeffmahoney committed:
go-libaudit: allow audit fd to be pollable

go-libaudit allows Receive to be called in non-blocking mode but doesn't
export the file descriptor to make it pollable.

jeffmahoney committed:
audit: use epoll to determine when messages are available

We use non-blocking mode in Receive to ensure that we don't end up
waiting forever when we intend to shut down, but the 1 second sleep to
wait for more events was a bandaid.  Now that libaudit exports its
file descriptor, we can use epoll to wait until events are available,
receive them all at once, and return to waiting.  This means that on
systems producing a lot of events, we'll be able to keep up and on
systems with few events we won't be retrying constantly.

jeffmahoney commented:
The updated PR that starts with commit 77798fe addresses most of the review from above. The updated 77798fe commit only contains some minor things: added some debugging messages, fixed minor bugs, cleaned up whitespace.  The "meat" of the update is in the final three commits which enable polling of the audit fd and ultimately switch it to unicast (which means it's also auditd). This allows us to keep up with thousands of messages/second without hitting ENOBUFS or dropping events. It does _not_ also log to `/var/log/audit/audit.log` and that's something we'll need to sort out. It may well be that what we want there is to accept events via auditd plugin, but I really want to get this into shape for broader testing.

jeffmahoney committed:
audit: implement buffer scaling

When a netlink socket returns ENOBUFS, it's expected that the caller
will scale the receive buffer via setsockopt appropriately.  We also
start with a 4 MB buffer to leave plenty of headroom.

jeffmahoney committed:
third_party/go-libaudit: move buffer handling from netlink to audit

In preparation to let the user of the audit API handle the receive
buffer, this commit moves the handling of the send/receive from the
netlink implementation to the caller.
There are no externally visible changes to the audit API.

jeffmahoney committed:
third_party/go-libaudit: move handling of receive buffer to caller

jeffmahoney committed:
audit: use lockless queue to accept messages

Processing the messages in the same thread that is responsible for
receiving them introduces enough latency that messages are getting
dropped.  This commit adds a lockless queue to hold the received
messages as soon as they are received and moves the processing of the
messages into a different thread.  This requires that we maintain a
separate receive buffer for each message received, so we use a pool
of buffers to avoid unnecessary memory alloc/release churn.

jeffmahoney committed:
audit: add statistics tracking

jeffmahoney committed:
audit: use 512k bufsize

jeffmahoney committed:
audit: fix shutdown sequence

lorddoskias commented:
LGTM

jeffmahoney committed:
audit: add helper to create stopped timer

jeffmahoney committed:
audit: improve scalability of receive/parse path

There are a lot of implicit allocations and copies in the receive/parse
path that just add overhead.  This commit makes a few changes to improve
that:

1) Don't create a new string using values that are already known just to
   have ParseLogLine pull them back out again.
2) Don't convert back and forth between []bytes and string unnecessarily
3) Don't allocate a new AuditMessage for every coalesced event
4) Don't copy bytes around when parsing the netlink packet when we can
   just access the parts we need directly.
5) Add an Owner field to AuditMessage that we can use to free the buffer
   from the ReassemblyComplete callback, which means that the same
   structure and memory  is used from initial packet receipt to final
   processing.

jeffmahoney committed:
audit: reindent listenerEventLoop

jeffmahoney committed:
audit: make event loop nonblocking

jeffmahoney committed:
audit: queue events and process in batches

jeffmahoney committed:
audit: eliminate racing disconnects

When distributeEvent queues up watchers to disconnect, it does so in a
goroutine.  Until the watcher completes the disconnect, new events will
also queue up the same watcher, leading to disconnect being called
multiple times from separate threads.

jeffmahoney committed:
audit: fix regression in Netlink.Send causing spurious ACK failures

jeffmahoney committed:
audit: allow disabling debug printing

lorddoskias commented on vql/linux/audit.go:
```diff
@@ -501,6 +521,26 @@ func (self *AuditWatcherService) startEventQueue() {
+	defer self.wgDec("startEventQueueMaintainer")
+
+	timeout := time.NewTicker(500 * time.Millisecond)
+	defer timeout.Stop()
```
Does this have any effect? You create the ticker but I don't see where you read from it? In the loop below you read from queueFlushTimer ?

jeffmahoney replied:
Thanks, yep. That's a leftover from before I had factored out the flushing and the timer. It should be removed.

jeffmahoney committed:
audot: (cleanup) remove timer left over from adding queue flush timer

jeffmahoney commented:
This has been obsoleted by PR#40. Thanks for the review and feedback, everyone!
